client/core/core.go

```diff
@@ -7406,3 +7439,233 @@ func (c *Core) findActiveOrder(oid order.OrderID) (*trackedTrade, error) {
+func (core *Core) fetchAllAssetRates() map[uint32]float64 {
+	supportedAssets := core.SupportedAssets()
+	assetRatesMap := make(map[uint32]float64, len(supportedAssets))
+	for assetID := range supportedAssets {
+		var rateSum float64
+		var sources int32
+		for _, fetcher := range core.exchangeRateFetchers {
+			rateInfo, found := fetcher.AssetRate(assetID)
+			if found && time.Since(rateInfo.lastUpdate) < fiatRateDataExpiry {
+				sources++
+				rateSum += rateInfo.rate
+			}
+		}
+		if rateSum != 0 {
+			assetRatesMap[assetID] = rateSum / float64(sources) // get average rate.
+		}
+	}
+
+	// Clean exchange rate fetchers with expired rate data.
+	for token, fetcher := range core.exchangeRateFetchers {
+		var expiredCount int
+		assetRates := fetcher.AssetRates()
+		for _, rateInfo := range assetRates {
+			if time.Since(rateInfo.lastUpdate) > fiatRateDataExpiry {
+				expiredCount++
+			}
+		}
+		// Check if all exchange rate data from rate fetcher are expired.
+		lenAssetRates := len(assetRates)
+		if expiredCount == lenAssetRates && lenAssetRates != 0 {
+			delete(core.exchangeRateFetchers, token)
+		}
+	}
+
+	// Save disabled exchange rate fetcher to database.
+	core.saveDisabledRateFetchers()
+
+	return assetRatesMap
+}
+
+// IsConversionDisabled checks if fiat rate fetch for assets is disabled.
+// This is when either no exchange rate fetcher is enabled or rates are
+// not up-to-date. Individual exchange rate fetcher can be disabled without
+// disabling rate fetching, as long as there is at least one exchange
+// rate fetcher.
+func (core *Core) IsConversionDisabled() bool {
+	core.ratesMtx.Lock()
+	defer core.ratesMtx.Unlock()
+	return len(core.exchangeRateFetchers) == 0
+}
+
+// ToggleExchangeRateFetcherStatus toggles an exchange rate fetcher status. If
+// disable is true, the rate fetcher (source) is disabled, otherwise the rate
+// fetcher is enabled.
+func (core *Core) ToggleExchangeRateFetcherStatus(source string, disable bool) error {
+	if disable {
+		return core.disableExchangeRateFetcher(source)
+	} else {
+		return core.enableExchangeRateFetcher(source)
+	}
+}
+
+// enableExchangeRateFetcher enables an exchange rate fetcher.
+func (core *Core) enableExchangeRateFetcher(source string) error {
+	core.ratesMtx.Lock()
+	defer core.ratesMtx.Unlock()
+
+	// Check if it's an invalid rate fetcher or it is already enabled.
+	fetcherConstructor := exchangeRateFetchers[source]
+	if fetcherConstructor == nil {
+		return fmt.Errorf("cannot enable unkown exchange rate fetcher")
+	} else if core.exchangeRateFetchers[source] != nil {
+		return nil // already enabled.
+	}
+
+	// Build exchange rate fetcher.
+	core.exchangeRateFetchers[source] = fetcherConstructor(core.log)
+
+	// If this is our first rate fetcher, start exchange rate fetcher
+	// gorountine, else fetch rates.
+	if len(core.exchangeRateFetchers) == 1 {
+		core.fetchExchangeRates()
+	} else {
+		go core.exchangeRateFetchers[source].FetchRates(core.assetMap())
+	}
```

Comment:
> Which is maybe supposed to stop the loop there, but could a fetcher be removed and added in the interval so that the length of the fetchers is never 0 for the already running core.fetchExchangeRates()?
Maybe simplest to just run the one loop regardless and for the life of core.


Yes, this allows users to toggle the rate fetchers as they deem fit without shutting down dexc. Also, we don't want to keep running a goroutine that does nothing?, assuming the user disabled all rate fetchers on the fly and then wants to enable rate fetching after some time (could be a very long time)?.

Reply:
It's my opinion that simplest is always best to reduce bugs. Not only bugs now, but bugs later when someone else touches the code.

This looks like it has a bug to me but I'm not entirely sure as I have not tested it irl.

I don't think that a loop that sometimes returns immediately every ten seconds is that bad.

Turning on/off the loop is also fine imo, as long as it works as intended.

Reply:
@JoeGruffins, thanks for reviewing.